components/tidb_query_datatype/src/codec/datum.rs

```diff
@@ -51,6 +51,72 @@ pub enum Datum {
     Max,
 }
 
+impl Datum {
+    #[inline]
+    pub fn as_int(&self) -> Result<Option<i64>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::I64(i) => Ok(Some(i)),
+            Datum::U64(u) => Ok(Some(u as i64)),
+            _ => Err(box_err!("Can't eval_int from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_real(&self) -> Result<Option<f64>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::F64(f) => Ok(Some(f)),
+            _ => Err(box_err!("Can't eval_real from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_decimal(&self) -> Result<Option<Cow<'_, Decimal>>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::Dec(ref d) => Ok(Some(Cow::Borrowed(d))),
+            _ => Err(box_err!("Can't eval_decimal from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_string(&self) -> Result<Option<Cow<'_, [u8]>>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::Bytes(ref b) => Ok(Some(Cow::Borrowed(b))),
+            _ => Err(box_err!("Can't eval_string from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_time(&self) -> Result<Option<Cow<'_, Time>>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::Time(t) => Ok(Some(Cow::Owned(t))),
+            _ => Err(box_err!("Can't eval_time from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_duration(&self) -> Result<Option<Duration>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::Dur(d) => Ok(Some(d)),
+            _ => Err(box_err!("Can't eval_duration from Datum")),
+        }
+    }
+
+    #[inline]
+    pub fn as_json(&self) -> Result<Option<Cow<'_, Json>>> {
+        match *self {
+            Datum::Null => Ok(None),
+            Datum::Json(ref j) => Ok(Some(Cow::Borrowed(j))),
+            _ => Err(box_err!("Can't eval_json from Datum")),
+        }
+    }
+}
+
```

Comment:
Where do they come from?

Reply:
Orphan rule prevent from compiling the code, struct `Datum` and `impl` belong the same crate before, but not after.